lib/pdf/reader/standard_security_handler.rb

```diff
@@ -92,7 +107,20 @@ def decrypt( buf, ref )
       rc4.decrypt(buf)
     end
 
-    private
+    # decrypt with AES-128-CBC algorithm
+    # when (version == 4 and CFM == AESV2)
+    def decrypt_aes128( buf, ref )
+      objKey = @encrypt_key.dup
+      (0..2).each { |e| objKey << (ref.id >> e*8 & 0xFF ) }
+      (0..1).each { |e| objKey << (ref.gen >> e*8 & 0xFF ) }
+      objKey << 'sAlT'  # Algorithm 1, b)
+      length = objKey.length < 16 ? objKey.length : 16
+      cipher = OpenSSL::Cipher.new("AES-#{length << 3}-CBC")
+      cipher.decrypt
+      cipher.key = Digest::MD5.digest(objKey)[0,length]
+      cipher.iv = buf[0..15]
+      cipher.update(buf[16..-1]) + cipher.final
+    end
```

Comment:
Thanks for splitting this out.

It's interesting that the different encryption algorithms use the same key construction process - the difference between decrypting an rc4 PDF and an AESV2 is so small....